vpr/src/place/cut_spreader.h

```diff
@@ -0,0 +1,259 @@
+ *
+ * find_overused_regions & expand_regions
+ * --------------------------------------
+ * @see find_overused_regions()
+ * @see expand_regions()
+ * The first step is to find an area of the FPGA that is overutilized for which the blocks contained within
+ * must be spread to a larger area. To obtain this overutilized area, adjacent locations on the FPGA that are
+ * occupied by more than one block (also overutilized) are repeatedly clustered together, until all clusters
+ * are bordered on all sides by non-overutilized locations. Next, the area is expanded in both the x and y
+ * directions until it's large enough to accommodate all blocks contained. Overutilization is defined as follows:
+ * (Occupancy / Capacity) > beta, where beta is a constant <=1, currently defined in AnalyticPlacer::PlacerHeapCfg.
+ *
+ * cut_region
+ * ----------
+ * @see cut_region()
+ * @see run()
+ * In the second step, two cuts are generated: a source cut and a target cut. The source cut pertains to the blocks
+ * being places; the target cut pertains to the area into which the blocks are placed. The source cut splits the
+ * blocks into two partitions, while the target cut splits the area into two sub-areas, into which the blocks in
+ * each partition are spread. Two objectives are minimized during this process: the imbalance between the number of
+ * blocks in each partition, and the difference in the utilization (Occupancy / Capacity) of each subarea.
+ *
+ * To generate the source cut, the logic blocks are sorted by their raw_x or raw_y location, depending on the
+ * orientation of the desired cut. After sorting, a pivot is chosen, all blocks to the left of the pivot are
+ * assigned to the left partition, and all blocks to the right are assigned to the right partition (we use left
+ * for left/top, right for right/bottom in x/y directions respectively in the implementation).
+ *
+ * The target cut is an x or y cut of the area such that all blocks in each partition fit in their respective
+ * subareas, and the difference in utilization is minimized. This is difficult due to the discrete nature of FPGA
+ * architecture. To achieve this, first assign all tiles to right region, then move tiles to the left one by one,
+ * calculating the difference in utilization for each move. The move with smallest delta_utilization is chosen as
+ * the target cut. To eliminate possible overutilization in either subarea, perturb the source cut by moving a
+ * single block from the over-utilized sub-area to the other, until neither is overutilized.
+ *
+ * Next, the blocks in sub-areas are spread to distribute them evenly. We split the sub-area into 10 equally-sized
+ * bins and logic blocks into 10 equal-capacity source bins. Then linearly interpolate to map blocks from their
+ * original_locations in their source bins to new spread location in target bins.
+ *
+ * This cut and spreading is repeated recursively. The cut spreading process returns the left and right subareas,
```

Comment:
Do you alternate x and y dimensions, with one x-area (e.g. left) being cut into two y-areas, and then each of those two y-subareas being cut into two further x-subareas?